testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java

```diff
@@ -1171,6 +1172,30 @@ public void testRenameColumn()
         super.testRenameColumn();
     }
 
+    @Test
+    public void testInsertIntoNotNullColumn()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE) && hasBehavior(SUPPORTS_INSERT_NOT_NULL_COLUMN));
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "insert_not_null", "(nullable_col INTEGER, not_null_col INTEGER NOT NULL)")) {
+            assertUpdate(format("INSERT INTO %s (not_null_col) VALUES (2)", table.getName()), 1);
+            assertQuery("SELECT * FROM " + table.getName(), "VALUES (NULL, 2)");
+            assertQueryFails(format("INSERT INTO %s (nullable_col) VALUES (1)", table.getName()), errorMessageForInsertIntoNotNullColumn("not_null_col"));
```

Comment:
Looks like a bug in implementation from https://github.com/trinodb/trino/pull/4144. I'd expect the engine to prevent inserts in both cases.

Is the reason that we don't create ColumnMetadata for all columns, only the columns that are part of the query?

Reply:
I don't think this is a bug (at least for now) because we have `supportsMissingColumnsOnInsert=true` in DefaultJdbcMetadata.